File: 1458-max-dot-product-of-two-subsequences/solution.py
class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        n, m = len(nums1), len(nums2)
        prev = [float("-inf")] * (m + 1)
        for i in range(1, n + 1):
            dp = [float("-inf")] * (m + 1)
            for j in range(1, m + 1):
                product = nums1[i-1] * nums2[j-1]
                term1 = product + max(0, prev[j-1])
                term2 = prev[j]
                term3 = dp[j-1]
                dp[j] = max(term1, term2, term3)
            prev = dp
        return int(prev[m])

File: 1458-max-dot-product-of-two-subsequences/solution.go
func maxDotProduct(nums1 []int, nums2 []int) int {
	m := len(nums2)
	prev := make([]int, m+1)

	const minVal = math.MinInt32

	for i := range prev {
		prev[i] = minVal
	}

	for _, num1 := range nums1 {
		dp := make([]int, m+1)
		for k := range dp {
			dp[k] = minVal
		}

		for j, num2 := range nums2 {
			product := num1 * num2
			col := j + 1

			prevVal := prev[col-1]
			if prevVal < 0 {
				prevVal = 0
			}
			term1 := product + prevVal
			term2 := prev[col]
			term3 := dp[col-1]
			dp[col] = max(term1, max(term2, term3))
		}
		prev = dp
	}

	return prev[m]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
